### ISSUE TITLE
Feature-segment

### ISSUE THREAD

**Issue by georgebuzzell** [NDCLab/baseEEG#27]
Feature-segment
-segment/epoch (cut) the continuous data into epochs of data, such that the zero point for each epoch is a given marker of interest
-write to annotations file which markers were used for epoching purposes, how many of each epoch were created, and how many ms appear before/after the markers of interest

**Comment by Jonhas** [comment]
@georgebuzzell  I began working on this, and I wanted to ask if I am on the right track. This is all so new to me so I want to assure I am understanding properly. Consider the following script. 
```python
#!/usr/bin/python3

import mne
from pathlib import Path
from mne_bids import BIDSPath, read_raw_bids

# Read BIDS data
bids_root = Path('eeg_matchingpennies')
bids_path = BIDSPath(subject='05',
                     task='matchingpennies',
                     datatype='eeg',
                     root=bids_root)
raw = read_raw_bids(bids_path)

# extract events and event_id from our raw object
events, event_id = mne.events_from_annotations(raw)

# we can go ahead and plot the events
mne.viz.plot_events(events, event_id=event_id)

# create an epoch object and we can extract information off of it
epochs = mne.Epochs(raw, events, event_id=event_id, preload=True)
epochs.plot(n_epochs=10)
print(epochs.info)
print(epochs[['left','right']])
```
When reading in BIDSPath, there are no issues with exception to missing coordinate system. We get the following output:
```vim 
Extracting parameters from eeg_matchingpennies/sub-05/eeg/sub-05_task-matchingpennies_eeg.vhdr...
Setting channel info structure...
Reading events from eeg_matchingpennies/sub-05/eeg/sub-05_task-matchingpennies_events.tsv.
Reading channel info from eeg_matchingpennies/sub-05/eeg/sub-05_task-matchingpennies_channels.tsv.
/home/ubuntu/.local/lib/python3.8/site-packages/mne_bids/path.py:1065: RuntimeWarning: Did not find any coordsystem.json associated with sub-05_task-matchingpennies.

The search_str was "eeg_matchingpennies/sub-05/**/sub-05*coordsystem.json"
  warn(msg)
/home/ubuntu/.local/lib/python3.8/site-packages/mne_bids/path.py:1065: RuntimeWarning: Did not find any eeg.json associated with sub-05_task-matchingpennies.

The search_str was "eeg_matchingpennies/sub-05/**/sub-05*eeg.json"
  warn(msg)
```
Then, I go on to extract events and event id from our raw object. So after running 
```python 
events, event_id = mne.events_from_annotations(raw)
```
I get the following output 
```vim 
Used Annotations descriptions: ['left', 'right']
```

I then go on to plot the events using 
```python3
mne.viz.plot_events(events, event_id = event_id)
```
resulting in 
![image](https://user-images.githubusercontent.com/45021859/114805129-0faba980-9d68-11eb-9a91-eadc288a225c.png)

I then go on to extract epochs from our data the following lines: 
```python3
epochs = mne.Epochs(raw, events, event_id=event_id, preload=True)
```
Now, when extracting information off of it, such as printing the epoch.info or print out the amount of epochs, we get 
```
<Info | 9 non-empty values
 bads: 2 items (FC5, FC6)
 ch_names: FC5, FC1, C3, CP5, CP1, FC2, FC6, C4, CP2, CP6
 chs: 10 EEG
 custom_ref_applied: False
 highpass: 0.0 Hz
 lowpass: 2500.0 Hz
 meas_date: unspecified
 nchan: 10
 projs: []
 sfreq: 5000.0 Hz
 subject_info: 4 items (dict)
>
900
```
With the following line, 
```python3 
print(epochs[['left','right']])
```
we get the following information back
```python3
<Epochs |  900 events (all good), -0.2 - 0.5 sec, baseline [-0.2, 0] sec, ~240.4 MB, data loaded,
 'left': 537
 'right': 363>
```
where -0.2 and 0.5 are the start and end time of the epochs in seconds, respectively. Would this be considered the seconds that appear before/after the markers of interest? and if so, does this abide to what we want to accomplish? To note, this script was converted to a jupyter notebook as described in #21

**Comment by georgebuzzell** [comment]
@Jonhas regaurding the warning you get, I got the same thing when testing another feature that @yanbin-niu is working on; see here for the discussion: https://github.com/NDCLab/baseEEG/issues/24 Unfortunately, it is not clear to me yet why we get this warning, given that there is a task-level _eeg.json file that SHOULD be applied to all data.  I am not currently sure if that is actually an issue or not. Please let us know if you have any thoughts on this. Regardless, I don't think this should be an issue for the current feature you are working on.

For the remainder of your script, this all looks correct to me. I have not directly tested your code yet, but if you are asking if you are on the right track and things generally look correct, then the answer is "yes"! :)

For the specific question about -.2 and .5, yes, you are correct that those are the amount of time (in seconds) before/after the marker zero point. Perfect! Great work! :) What you will now want to do is modify the code slightly such that there are variables in your code for how much time before and after to cut. Then, these variables can be read in by a user-supplied json file. That way, when the preprocessing script is run in the future, a user need only edit fields in the json file and it will change how various parameters of the preprocessing features perform. Does that make sense? Relatedly, I am going to put discussion of this user supplied json file on the agenda for the next meeting, as I had not mentioned it previously.

Again, great work, @Jonhas! Thank you!!

**Comment by F-said** [comment]
closing per merged PR #97
